File: post-order-traversal/post-order.ml
(* Binary tree data type: Empty or Node(left, value, right) *)
datatype 'data tree = Empty | Node of 'data tree * 'data * 'data tree;

(* Post-order traversal: left @ right @ [current] *)
fun postorder Empty = []  (* Empty tree returns empty list *)
  | postorder (Node (left, value, right)) =  (* Pattern match on node *)
      postorder left @      (* Traverse left subtree *)
      postorder right @     (* Traverse right subtree *)
      [value];              (* Add current node value *)

(* Example tree: same structure as other implementations *)
val myTree = Node(
    Node(Node(Empty, 4, Empty), 2, Node(Empty, 5, Empty)),  (* Left subtree *)
    1,                                                       (* Root *)
    Node(Empty, 3, Node(Empty, 6, Empty))                   (* Right subtree *)
);

(* Execute traversal - returns [4, 5, 2, 6, 3, 1] *)
postorder myTree;

File: post-order-traversal/post-order.pl
% Post Order Traversal in Prolog

% Base case: empty tree returns empty list
postorder(nil, []).

% Recursive case: traverse left, right, then current node
postorder(node(Left, Value, Right), Result) :-
    postorder(Left, LeftResult),        % Get left subtree result
    postorder(Right, RightResult),      % Get right subtree result
    append(LeftResult, RightResult, TempResult),  % Combine left and right
    append(TempResult, [Value], Result).          % Add current node value

% Example tree structure: node(Left, Value, Right)
% Tree: 1 with children 2(4,5) and 3(6)
% Expected result: [4, 5, 2, 6, 3, 1]

% run this query to see result
% postorder(node(node(node(nil, 4, nil), 2, node(nil, 5, nil)), 1, node(nil, 3, node(nil, 6, nil))), Result).